### ISSUE TITLE
OKD install fails after bootstrap

### ISSUE THREAD

**Issue by fortinj66** [openshift/okd#373]
**Describe the bug**
OKD install fails on vSphere IPI after bootstrap completes.  I get the following:

```
INFO Waiting up to 30m0s for the cluster at https://api.dev-c1v4.os.maeagle.corp:6443 to initialize...
DEBUG Still waiting for the cluster to initialize: Multiple errors are preventing progress:
* Could not update oauthclient "console" (363 of 596): the server does not recognize this resource, check extension API servers
* Could not update prometheusrule "openshift-cloud-credential-operator/cloud-credential-operator-alerts" (198 of 596): the server does not recognize this resource, check extension API servers
* Could not update prometheusrule "openshift-cluster-samples-operator/samples-operator-alerts" (315 of 596): the server does not recognize this resource, check extension API servers
* Could not update servicemonitor "openshift-authentication-operator/authentication-operator" (510 of 596): the server does not recognize this resource, check extension API servers
* Could not update servicemonitor "openshift-cluster-machine-approver/cluster-machine-approver" (520 of 596): the server does not recognize this resource, check extension API servers
* Could not update servicemonitor "openshift-cluster-version/cluster-version-operator" (8 of 596): the server does not recognize this resource, check extension API servers
* Could not update servicemonitor "openshift-config-operator/config-operator" (124 of 596): the server does not recognize this resource, check extension API servers
* Could not update servicemonitor "openshift-image-registry/image-registry" (516 of 596): the server does not recognize this resource, check extension API servers
* Could not update servicemonitor "openshift-machine-api/cluster-autoscaler-operator" (230 of 596): the server does not recognize this resource, check extension API servers
DEBUG Still waiting for the cluster to initialize: Working towards 4.5.0-0.okd-2020-10-15-235428
DEBUG Still waiting for the cluster to initialize: Working towards 4.5.0-0.okd-2020-10-15-235428: downloading update
DEBUG Still waiting for the cluster to initialize: Working towards 4.5.0-0.okd-2020-10-15-235428
DEBUG Still waiting for the cluster to initialize: Working towards 4.5.0-0.okd-2020-10-15-235428: 1% complete
DEBUG Still waiting for the cluster to initialize: Working towards 4.5.0-0.okd-2020-10-15-235428: 3% complete
DEBUG Still waiting for the cluster to initialize: Working towards 4.5.0-0.okd-2020-10-15-235428: 10% complete
DEBUG Still waiting for the cluster to initialize: Working towards 4.5.0-0.okd-2020-10-15-235428: 13% complete
DEBUG Still waiting for the cluster to initialize: Working towards 4.5.0-0.okd-2020-10-15-235428: 49% complete
DEBUG Still waiting for the cluster to initialize: Working towards 4.5.0-0.okd-2020-10-15-235428: 62% complete
DEBUG Still waiting for the cluster to initialize: Working towards 4.5.0-0.okd-2020-10-15-235428: 82% complete
DEBUG Still waiting for the cluster to initialize: Working towards 4.5.0-0.okd-2020-10-15-235428: 83% complete
DEBUG Still waiting for the cluster to initialize: Working towards 4.5.0-0.okd-2020-10-15-235428: 85% complete
DEBUG Still waiting for the cluster to initialize: Working towards 4.5.0-0.okd-2020-10-15-235428: 86% complete, waiting on authentication, cluster-autoscaler, console, ingress, kube-storage-version-migrator, machine-api, monitoring
DEBUG Still waiting for the cluster to initialize: Working towards 4.5.0-0.okd-2020-10-15-235428: 86% complete
DEBUG Still waiting for the cluster to initialize: Working towards 4.5.0-0.okd-2020-10-15-235428: 86% complete, waiting on authentication, console, ingress, kube-storage-version-migrator, monitoring
ERROR Cluster operator authentication Degraded is True with IngressStateEndpoints_MissingEndpoints::RouteStatus_FailedHost: IngressStateEndpointsDegraded: No endpoints found for oauth-server
RouteStatusDegraded: route is not available at canonical host oauth-openshift.apps.dev-c1v4.os.maeagle.corp: []
INFO Cluster operator authentication Progressing is Unknown with NoData:
INFO Cluster operator authentication Available is Unknown with NoData:
INFO Cluster operator console Progressing is True with DefaultRouteSync_FailedAdmitDefaultRoute::OAuthClientSync_FailedHost: DefaultRouteSyncProgressing: route "console" is not available at canonical host
OAuthClientSyncProgressing: route "console" is not available at canonical host []
INFO Cluster operator console Available is Unknown with NoData:
INFO Cluster operator ingress Available is False with IngressUnavailable: Not all ingress controllers are available.
INFO Cluster operator ingress Progressing is True with Reconciling: Not all ingress controllers are available.
ERROR Cluster operator ingress Degraded is True with IngressControllersDegraded: Some ingresscontrollers are degraded: default
INFO Cluster operator insights Disabled is False with :
INFO Cluster operator kube-storage-version-migrator Available is False with _NoMigratorPod: Available: deployment/migrator.openshift-kube-storage-version-migrator: no replicas are available
ERROR Cluster operator monitoring Degraded is True with UpdatingAlertmanagerFailed: Failed to rollout the stack. Error: running task Updating Alertmanager failed: waiting for Alertmanager Route to become dy failed: waiting for route openshift-monitoring/alertmanager-main: no status available
INFO Cluster operator monitoring Available is False with :
INFO Cluster operator monitoring Progressing is True with RollOutInProgress: Rolling out the stack.
FATAL failed to initialize the cluster: Some cluster operators are still updating: authentication, console, ingress, kube-storage-version-migrator, monitoring

Note that the worker nodes are created, but are not started
```

**Version**
vSphere IPI
./openshift-install 4.5.0-0.okd-2020-10-15-235428
built from commit 63200c80c431b8dbaa06c0cc13282d819bd7e5f8
release image quay.io/openshift/okd@sha256:67cc7cb47d56237adcf0ecc2ee76446785add5fa236cd08746b55f578b9200a5

**How reproducible**
currently 100% reproducible

**Log bundle**
Not quite sure what logs to capture here as the cluster is not running and bootstrap node has been destroyed...

**Note that I had had a working cluster prior to this install.  I deleted that cluster in order to solidify the install process.  Nothing in the infrastructure changed and I used the same install parameters.**

**Comment by vrutkovs** [comment]
>Not quite sure what logs to capture here as the cluster is not running and bootstrap node has been destroyed...

https://docs.okd.io/latest/support/gathering-cluster-data.html#support_gathering_data_gathering-cluster-data

**Comment by fortinj66** [comment]
Very odd...   I tried the install 5 times in a row with it failing as described, although the 5th time it only had one operator missing/waiting and I was able to log into the console.  In that case it seemed the worker node had started but memory constrained...  Nodes were set as the default 2CPU/8GMem.  I manually increased CPU and Memory to 4/24Gig and the install completed.

This morning I tore everything down and changed the default worker node settings to 4CPU and 24Gig ram and the install completed without issue:

```
compute:
- architecture: amd64
  hyperthreading: Enabled
  name: worker
  platform:
    vsphere:
      cpus: 4
      memoryMB: 24576
  replicas: 3

```

```
INFO Install complete!
INFO To access the cluster as the system:admin user when using 'oc', run 'export KUBECONFIG=/root/openshift-install-linux-4.5.0-0/okd-install-4.5/auth/kubeconfig'
INFO Access the OpenShift web-console here: https://console-openshift-console.apps.dev-c1v4.os.xxxxx.xxxxx
INFO Login to the console with user: "kubeadmin", and password: "nRkf8-cjYzS-vQiRA-ouU6L"
DEBUG Time elapsed per stage:
DEBUG     Infrastructure: 44s
DEBUG Bootstrap Complete: 13m56s
DEBUG                API: 3m56s
DEBUG  Bootstrap Destroy: 15s
DEBUG  Cluster Operators: 17m27s
INFO Time elapsed: 32m31s
```

So maybe the default worker node config (2CPU/8Gig) is too small?

**Comment by vrutkovs** [comment]
4 vCPUs and 16 GB RAM for masters is minimum. Not sure about workers - you might need 4vCPUs. 8GB RAM should be sufficient though.

must-gather would show the state of the pods and node utilization

**Comment by fortinj66** [comment]
The masters seem fine...  The worker pressure was on Memory.  I'm testing again with the defaults...

**Comment by fortinj66** [comment]
![image](https://user-images.githubusercontent.com/8594283/98859162-bdb8cd80-242f-11eb-8df9-64ba94bfdb1d.png)

openshift-kube-storage-version-migrator is failing because the node doesn't have enough memory available:

Here are the node resources:

```
Allocated resources:
  (Total limits may be over 100 percent, i.e., overcommitted.)
  Resource           Requests      Limits
  --------           --------      ------
  cpu                1083m (72%)   0 (0%)
  memory             6678Mi (97%)  512Mi (7%)
  ephemeral-storage  0 (0%)        0 (0%)
  hugepages-1Gi      0 (0%)        0 (0%)
  hugepages-2Mi      0 (0%)        0 (0%)

```

Note that Memory is at 97% requested  (8 Gig on the node)


What is odd is that the other two worker nodes did not get started automatically...  I started the second worker node and the dev-c1v4-cwvhn-worker-m5rlb pod was assigned to it and the cluster install completed successfully...

**Comment by vrutkovs** [comment]
>I can't run must-gather as the bootstrap process finishes successfully

must-gather doesn't require bootstrap

**Comment by fortinj66** [comment]
My bad, I was getting the bootstrap log gather confused with must-gather.

I am rebuilding the cluster (again lol) and I'll capture the logs.

**Comment by mschep** [comment]
I've got very similar results at the moment, however for me it is the first time I'm setting up OKD. 

I've run the installer with the defaults, so without a install-config.yaml, but also with one, and set the workers to have 16gb mem. The results are the same, the install fails and I've pasted below.

After the install fails there are just the 3 masters running in vSphere. There are no other nodes running, just those 3 masters. 

`INFO Waiting up to 30m0s for the cluster at https://api.okd.example.com:6443 to initialize... 
DEBUG Still waiting for the cluster to initialize: Working towards 4.5.0-0.okd-2020-10-15-235428: 67% complete 
DEBUG Still waiting for the cluster to initialize: Working towards 4.5.0-0.okd-2020-10-15-235428 
DEBUG Still waiting for the cluster to initialize: Working towards 4.5.0-0.okd-2020-10-15-235428: downloading update 
DEBUG Still waiting for the cluster to initialize: Working towards 4.5.0-0.okd-2020-10-15-235428 
DEBUG Still waiting for the cluster to initialize: Working towards 4.5.0-0.okd-2020-10-15-235428: 5% complete 
DEBUG Still waiting for the cluster to initialize: Working towards 4.5.0-0.okd-2020-10-15-235428: 7% complete 
DEBUG Still waiting for the cluster to initialize: Working towards 4.5.0-0.okd-2020-10-15-235428: 8% complete 
DEBUG Still waiting for the cluster to initialize: Working towards 4.5.0-0.okd-2020-10-15-235428: 50% complete 
DEBUG Still waiting for the cluster to initialize: Working towards 4.5.0-0.okd-2020-10-15-235428: 65% complete 
DEBUG Still waiting for the cluster to initialize: Working towards 4.5.0-0.okd-2020-10-15-235428: 74% complete 
DEBUG Still waiting for the cluster to initialize: Working towards 4.5.0-0.okd-2020-10-15-235428: 83% complete 
DEBUG Still waiting for the cluster to initialize: Working towards 4.5.0-0.okd-2020-10-15-235428: 85% complete 
DEBUG Still waiting for the cluster to initialize: Working towards 4.5.0-0.okd-2020-10-15-235428: 86% complete, waiting on authentication, cluster-autoscaler, console, ingress, kube-storage-version-migrator, machine-api, monitoring 
DEBUG Still waiting for the cluster to initialize: Working towards 4.5.0-0.okd-2020-10-15-235428: 86% complete 
DEBUG Still waiting for the cluster to initialize: Working towards 4.5.0-0.okd-2020-10-15-235428: 86% complete, waiting on authentication, console, ingress, kube-storage-version-migrator, monitoring 
DEBUG Still waiting for the cluster to initialize: Working towards 4.5.0-0.okd-2020-10-15-235428: 86% complete, waiting on authentication, console, ingress, kube-storage-version-migrator, monitoring 
DEBUG Still waiting for the cluster to initialize: Working towards 4.5.0-0.okd-2020-10-15-235428 
DEBUG Still waiting for the cluster to initialize: Working towards 4.5.0-0.okd-2020-10-15-235428: downloading update 
DEBUG Still waiting for the cluster to initialize: Working towards 4.5.0-0.okd-2020-10-15-235428 
DEBUG Still waiting for the cluster to initialize: Working towards 4.5.0-0.okd-2020-10-15-235428: 3% complete 
DEBUG Still waiting for the cluster to initialize: Working towards 4.5.0-0.okd-2020-10-15-235428: 8% complete 
DEBUG Still waiting for the cluster to initialize: Working towards 4.5.0-0.okd-2020-10-15-235428: 9% complete 
DEBUG Still waiting for the cluster to initialize: Working towards 4.5.0-0.okd-2020-10-15-235428: 85% complete 
DEBUG Still waiting for the cluster to initialize: Working towards 4.5.0-0.okd-2020-10-15-235428: 86% complete 
DEBUG Still waiting for the cluster to initialize: Working towards 4.5.0-0.okd-2020-10-15-235428: 86% complete, waiting on authentication, console, ingress, kube-storage-version-migrator, monitoring 
DEBUG Still waiting for the cluster to initialize: Some cluster operators are still updating: authentication, console, ingress, kube-storage-version-migrator, monitoring 
ERROR Cluster operator authentication Degraded is True with IngressStateEndpoints_MissingEndpoints::RouteStatus_FailedHost: IngressStateEndpointsDegraded: No endpoints found for oauth-server
RouteStatusDegraded: route is not available at canonical host oauth-openshift.apps.okd.example.com: [] 
INFO Cluster operator authentication Progressing is Unknown with NoData:  
INFO Cluster operator authentication Available is Unknown with NoData:  
INFO Cluster operator console Progressing is True with DefaultRouteSync_FailedAdmitDefaultRoute::OAuthClientSync_FailedHost: DefaultRouteSyncProgressing: route "console" is not available at canonical host []
OAuthClientSyncProgressing: route "console" is not available at canonical host [] 
INFO Cluster operator console Available is Unknown with NoData:  
INFO Cluster operator ingress Available is False with IngressUnavailable: Not all ingress controllers are available. 
INFO Cluster operator ingress Progressing is True with Reconciling: Not all ingress controllers are available. 
ERROR Cluster operator ingress Degraded is True with IngressControllersDegraded: Some ingresscontrollers are degraded: default 
INFO Cluster operator insights Disabled is False with :  
INFO Cluster operator kube-storage-version-migrator Available is False with _NoMigratorPod: Available: deployment/migrator.openshift-kube-storage-version-migrator: no replicas are available 
INFO Cluster operator monitoring Progressing is True with RollOutInProgress: Rolling out the stack. 
ERROR Cluster operator monitoring Degraded is True with UpdatingAlertmanagerFailed: Failed to rollout the stack. Error: running task Updating Alertmanager failed: waiting for Alertmanager Route to become ready failed: waiting for route openshift-monitoring/alertmanager-main: no status available 
INFO Cluster operator monitoring Available is False with :  
FATAL failed to initialize the cluster: Some cluster operators are still updating: authentication, console, ingress, kube-storage-version-migrator, monitoring `

**Comment by mschep** [comment]
One thing I noticed on the masters, is that their /etc/resolv.conf contains 3 nameservers (in my case):

```
[core@okd-2m4s4-master-2 etc]$ cat resolv.conf
# Generated by KNI resolv prepender NM dispatcher script
search okd.example.com
nameserver  xxx.0.4.246 <- local resolver on the master
nameserver 192.168.2.1 <- resolver gotten from DHCP request
nameserver 2001:xxx:5::1 <- resolver gotten from router advertisements 
```

Looking at the logs I see the masters are also trying to connect to api-int, which can only be resolved to by the local resolver and not the resolver gotten from DHCP and RA. 

`Nov 12 14:29:56.252598 okd-2m4s4-master-2 crio[29176]: 2020-11-12T14:29:56Z [error] SetNetworkStatus: failed to query the pod revision-pruner-3-okd-2m4s4-master-2 in out of cluster comm: Get https://[api-int.okd.example.com]:6443/api/v1/namespaces/openshift-etcd/pods/revision-pruner-3-okd-2m4s4-master-2?timeout=1m0s: dial tcp: lookup api-int.okd.cloud.ripe.net on 192.168.2.1:53: no such host`

Could this be the cause of the issue, and not so much the memory issue?

**Comment by fortinj66** [comment]
I have recreated the issue.  

Attached is the install log
[openshift_install.log](https://github.com/openshift/okd/files/5532546/openshift_install.log)

and a link to the must-gather data.  It was too big for here:  https://www.dropbox.com/s/pslai5un6dnbzz8/must-gather.local.2917834136669564543.tar.gz?dl=0

and current status of cluster ( over 2 hours later)

```
[root@os-utils-d02 openshift-install-linux-4.5.0-0]# date
Thu Nov 12 13:37:56 EST 2020
[root@os-utils-d02 openshift-install-linux-4.5.0-0]# ./openshift-install wait-for install-complete --log-level=debug --dir=okd-install-4.5
DEBUG OpenShift Installer 4.5.0-0.okd-2020-10-15-235428
DEBUG Built from commit 63200c80c431b8dbaa06c0cc13282d819bd7e5f8
DEBUG Fetching Install Config...
DEBUG Loading Install Config...
DEBUG   Loading SSH Key...
DEBUG   Loading Base Domain...
DEBUG     Loading Platform...
DEBUG   Loading Cluster Name...
DEBUG     Loading Base Domain...
DEBUG     Loading Platform...
DEBUG   Loading Pull Secret...
DEBUG   Loading Platform...
DEBUG Using Install Config loaded from state file
DEBUG Reusing previously-fetched Install Config
INFO Waiting up to 30m0s for the cluster at https://api.dev-c1v4.os.maeagle.corp:6443 to initialize...
DEBUG Still waiting for the cluster to initialize: Some cluster operators are still updating: authentication, console, ingress, kube-storage-version-migrator, monitoring
```

**Comment by vrutkovs** [comment]
>Looking at the logs I see the masters are also trying to connect to api-int, which can only be resolved to by the local resolver and not the resolver gotten from DHCP and RA.

Yeah, this seems to be the problem. Interesting, I don't see `mdns-publisher` pod in must-gather, but kubelet logs mention its running fine

**Comment by mschep** [comment]
> Yeah, this seems to be the problem. Interesting, I don't see mdns-publisher pod in must-gather, but kubelet logs mention its running fine.

Sorry, that's my issue and I've not uploaded must-gather (yet). It seems I am facing another issue than fortinj66, because his logs don't seem to show this resolver issue. Should I upload my must-gather or should I open another  issue?

**Comment by fortinj66** [comment]
[root@os-utils-d02 openshift-install-linux-4.5.0-0]# oc get pods
NAME                                     READY   STATUS    RESTARTS   AGE
coredns-dev-c1v4-2pdf5-master-0          2/2     Running   0          3h30m
coredns-dev-c1v4-2pdf5-master-1          2/2     Running   0          3h30m
coredns-dev-c1v4-2pdf5-master-2          2/2     Running   0          3h31m
haproxy-dev-c1v4-2pdf5-master-0          2/2     Running   4          3h31m
haproxy-dev-c1v4-2pdf5-master-1          2/2     Running   3          3h31m
haproxy-dev-c1v4-2pdf5-master-2          2/2     Running   2          3h31m
keepalived-dev-c1v4-2pdf5-master-0       2/2     Running   0          3h31m
keepalived-dev-c1v4-2pdf5-master-1       2/2     Running   0          3h30m
keepalived-dev-c1v4-2pdf5-master-2       2/2     Running   0          3h31m
mdns-publisher-dev-c1v4-2pdf5-master-0   1/1     Running   0          3h31m
mdns-publisher-dev-c1v4-2pdf5-master-1   1/1     Running   0          3h30m
mdns-publisher-dev-c1v4-2pdf5-master-2   1/1     Running   0          3h32m

**Comment by vrutkovs** [comment]
>Should I upload my must-gather or should I open another issue?

I think a new issue would be best

**Comment by fortinj66** [comment]
@vrutkovs what are my next steps?  

This isn't 100% reproducible but seems to be an issue...  I've had it happen probably 6/10 installs with the other 4 progressing but needing manual intervention with either an increase of memory or starting another worker as only one seems to start automatically...

**Comment by fortinj66** [comment]
Interesting...   I happened to look at the csr requests and saw the following:
```

[root@os-utils-d02 openshift-install-linux-4.5.0-0]# oc get nodes
NAME                      STATUS   ROLES    AGE    VERSION
dev-c1v4-2pdf5-master-0   Ready    master   4h5m   v1.18.3
dev-c1v4-2pdf5-master-1   Ready    master   4h5m   v1.18.3
dev-c1v4-2pdf5-master-2   Ready    master   4h6m   v1.18.3

[root@os-utils-d02 openshift-install-linux-4.5.0-0]# oc get csr
NAME        AGE   SIGNERNAME                                    REQUESTOR                                                                   CONDITION
csr-5rb7n   28m   kubernetes.io/kube-apiserver-client-kubelet   system:serviceaccount:openshift-machine-config-operator:node-bootstrapper   Pending
csr-7fr7l   12m   kubernetes.io/kube-apiserver-client-kubelet   system:serviceaccount:openshift-machine-config-operator:node-bootstrapper   Pending
csr-hkcvq   90m   kubernetes.io/kube-apiserver-client-kubelet   system:serviceaccount:openshift-machine-config-operator:node-bootstrapper   Pending
csr-k4w4j   59m   kubernetes.io/kube-apiserver-client-kubelet   system:serviceaccount:openshift-machine-config-operator:node-bootstrapper   Pending
csr-wnd7j   43m   kubernetes.io/kube-apiserver-client-kubelet   system:serviceaccount:openshift-machine-config-operator:node-bootstrapper   Pending
csr-x8dwl   74m   kubernetes.io/kube-apiserver-client-kubelet   system:serviceaccount:openshift-machine-config-operator:node-bootstrapper   Pending
[root@os-utils-d02 openshift-install-linux-4.5.0-0]# oc get csr -o go-template='{{range .items}}{{if not .status}}{{.metadata.name}}{{"\n"}}{{end}}{{end}}' | xargs oc adm certificate approve
certificatesigningrequest.certificates.k8s.io/csr-5rb7n approved
certificatesigningrequest.certificates.k8s.io/csr-7fr7l approved
certificatesigningrequest.certificates.k8s.io/csr-hkcvq approved
certificatesigningrequest.certificates.k8s.io/csr-k4w4j approved
certificatesigningrequest.certificates.k8s.io/csr-wnd7j approved
certificatesigningrequest.certificates.k8s.io/csr-x8dwl approved

[root@os-utils-d02 openshift-install-linux-4.5.0-0]# oc get csr
NAME        AGE   SIGNERNAME                                    REQUESTOR                                                                   CONDITION
csr-5rb7n   28m   kubernetes.io/kube-apiserver-client-kubelet   system:serviceaccount:openshift-machine-config-operator:node-bootstrapper   Approved,Issued
csr-7fr7l   13m   kubernetes.io/kube-apiserver-client-kubelet   system:serviceaccount:openshift-machine-config-operator:node-bootstrapper   Approved,Issued
csr-hkcvq   90m   kubernetes.io/kube-apiserver-client-kubelet   system:serviceaccount:openshift-machine-config-operator:node-bootstrapper   Approved,Issued
csr-k4w4j   59m   kubernetes.io/kube-apiserver-client-kubelet   system:serviceaccount:openshift-machine-config-operator:node-bootstrapper   Approved,Issued
csr-wnd7j   44m   kubernetes.io/kube-apiserver-client-kubelet   system:serviceaccount:openshift-machine-config-operator:node-bootstrapper   Approved,Issued
csr-x8dwl   75m   kubernetes.io/kube-apiserver-client-kubelet   system:serviceaccount:openshift-machine-config-operator:node-bootstrapper   Approved,Issued

[root@os-utils-d02 openshift-install-linux-4.5.0-0]# oc get nodes
NAME                          STATUS     ROLES    AGE    VERSION
dev-c1v4-2pdf5-master-0       Ready      master   4h7m   v1.18.3
dev-c1v4-2pdf5-master-1       Ready      master   4h7m   v1.18.3
dev-c1v4-2pdf5-master-2       Ready      master   4h8m   v1.18.3
dev-c1v4-2pdf5-worker-954tn   NotReady   worker   0s     v1.18.3
```

also saw a bunch of these in the events:

```0/3 nodes are available: 3 node(s) didn't match node selector.```

So I'm wondering if the CSRs not being approved caused the worker node not to be started?

**Comment by fortinj66** [comment]
After the first node started I got the following:

```INFO Waiting up to 30m0s for the cluster at https://api.dev-c1v4.os.maeagle.corp:6443 to initialize...
DEBUG Still waiting for the cluster to initialize: Cluster operator kube-storage-version-migrator has not yet reported success
```

That seems to indicate that the first node is out of memory.  

I then started a second worker node and saw the following:

```# oc get csr
NAME        AGE     SIGNERNAME                                    REQUESTOR                                                                   CONDITION
csr-5rb7n   52m     kubernetes.io/kube-apiserver-client-kubelet   system:serviceaccount:openshift-machine-config-operator:node-bootstrapper   Approved,Issued
csr-7fr7l   36m     kubernetes.io/kube-apiserver-client-kubelet   system:serviceaccount:openshift-machine-config-operator:node-bootstrapper   Approved,Issued
csr-flfmn   2m51s   kubernetes.io/kube-apiserver-client-kubelet   system:serviceaccount:openshift-machine-config-operator:node-bootstrapper   Pending
csr-g5kkm   23m     kubernetes.io/kubelet-serving                 system:node:dev-c1v4-2pdf5-worker-954tn                                     Approved,Issued
csr-hkcvq   114m    kubernetes.io/kube-apiserver-client-kubelet   system:serviceaccount:openshift-machine-config-operator:node-bootstrapper   Approved,Issued
csr-k4w4j   83m     kubernetes.io/kube-apiserver-client-kubelet   system:serviceaccount:openshift-machine-config-operator:node-bootstrapper   Approved,Issued
csr-wnd7j   67m     kubernetes.io/kube-apiserver-client-kubelet   system:serviceaccount:openshift-machine-config-operator:node-bootstrapper   Approved,Issued
csr-x8dwl   98m     kubernetes.io/kube-apiserver-client-kubelet   system:serviceaccount:openshift-machine-config-operator:node-bootstrapper   Approved,Issued
```

I waited about 5 minutes for the csr to be approved and it was not...  I manually approved it and then the node csr was created and approved automatically.

Once the worker node became ready the cluster install finished.

```INFO Install complete!
INFO To access the cluster as the system:admin user when using 'oc', run 'export KUBECONFIG=/root/openshift-install-linux-4.5.0-0/okd-install-4.5/auth/kubeconfig'
INFO Access the OpenShift web-console here: https://console-openshift-console.apps.dev-c1v4.os.maeagle.corp
INFO Login to the console with user: "kubeadmin", and password: "cK9mC-MXIEe-i6hgi-zptxm"
DEBUG Time elapsed per stage:
DEBUG Cluster Operators: 4m14s
INFO Time elapsed: 4m14s
```

So now it seems I can recover from this issue, but it seems to me there is an issue...  Maybe with CSR approval and worker node starts?  That's pretty much all I had to do to cause this to recover...

**Comment by vrutkovs** [comment]
machine-api should autoapprove CSRs from workers automatically. Attach must-gather for the case when it doesn't happen (also for api-int not being found)

**Comment by fortinj66** [comment]
The must-gather I provided earlier should have that information. It is the from the cluster I am currently testing on and had to approve the CSRs

**Comment by vrutkovs** [comment]
>The must-gather I provided earlier should have that information

It doesn't have logs from `openshift-vsphere-infra` namespace, could you check it again?

**Comment by fortinj66** [comment]
Here is the current must-gather:  https://www.dropbox.com/s/y7oemtf67tnmhbl/must-gather.local.4179956643971503215.tar.gz?dl=0

Note that this is not seemingly related to the api-int issue reported by @mschep

**Comment by fortinj66** [comment]
I started the third worker node and I see this:

```
 oc get csr
NAME        AGE     SIGNERNAME                                    REQUESTOR                                                                   CONDITION
csr-flfmn   82m     kubernetes.io/kube-apiserver-client-kubelet   system:serviceaccount:openshift-machine-config-operator:node-bootstrapper   Approved,Issued
csr-gxrht   78m     kubernetes.io/kubelet-serving                 system:node:dev-c1v4-2pdf5-worker-fxghk                                     Approved,Issued
csr-z7qfj   5m52s   kubernetes.io/kube-apiserver-client-kubelet   system:serviceaccount:openshift-machine-config-operator:node-bootstrapper   Pending
```
Note the pending CSR has been pending almost 6 minutes...

**Comment by vrutkovs** [comment]
(lol sorry, three avatars in sunglasses and late here so I got confused :) )

So I see 3 machines:
* dev-c1v4-2pdf5-worker-954tn
* dev-c1v4-2pdf5-worker-fxghk
* dev-c1v4-2pdf5-worker-nz6k7

and only two nodes:

* dev-c1v4-2pdf5-worker-954tn
* dev-c1v4-2pdf5-worker-fxghk

and two approved csrs - both for `fxghk` node.

Some CSRs have wrong date:
```
CSR csr-7fr7l not authorized: CSR csr-7fr7l creation time 2020-11-12 20:28:38 +0000 UTC not in range (2020-11-12 16:41:57 +0000 UTC, 2020-11-12 18:42:07 +0000 UTC)
```

Make sure your machines have the same timezone?

**Comment by fortinj66** [comment]
Some of those were probably from CSRs being regenerated...

Where did you see those entries?

**Comment by vrutkovs** [comment]
`./namespaces/openshift-cluster-machine-approver/pods/machine-approver-86756d9f66-hk5q5` is the pod which auto approves CSRs for known nodes.

Check that NTP is not blocked on your system and nodes boot with correct time. Not sure why its different across different worker nodes (and masters are probably not affected)

**Comment by fortinj66** [comment]
I'm seeing it:

```
I1112 22:37:31.815668       1 main.go:147] CSR csr-8wncm added
I1112 22:37:31.842708       1 main.go:182] CSR csr-8wncm not authorized: CSR csr-8wncm creation time 2020-11-12 22:34:47 +0000 UTC not in range (2020-11-12 16:41:57 +0000 UTC, 2020-11-12 18:42:07 +0000 UTC)
I1112 22:37:31.842742       1 main.go:218] Error syncing csr csr-8wncm: CSR csr-8wncm creation time 2020-11-12 22:34:47 +0000 UTC not in range (2020-11-12 16:41:57 +0000 UTC, 2020-11-12 18:42:07 +0000 UTC)
```
let me check the clients

**Comment by fortinj66** [comment]
The time on the CSR is correct.  I'm in the US EastCoast which is currently -0500.  I checked the master and worker nodes and they all have the correct time.

```
[root@os-utils-d02 ~]# oc describe csr csr-8wncm
Name:               csr-8wncm
Labels:             <none>
Annotations:        <none>
CreationTimestamp:  Thu, 12 Nov 2020 17:34:47 -0500
Requesting User:    system:serviceaccount:openshift-machine-config-operator:node-bootstrapper
Signer:             kubernetes.io/kube-apiserver-client-kubelet
Status:             Pending
Subject:
         Common Name:    system:node:dev-c1v4-2pdf5-worker-nz6k7
         Serial Number:
         Organization:   system:nodes
Events:  <none>
```
```I1112 22:56:38.779160       1 main.go:147] CSR csr-8wncm added
I1112 22:56:38.814409       1 main.go:182] CSR csr-8wncm not authorized: CSR csr-8wncm creation time 2020-11-12 22:34:47 +0000 UTC not in range (2020-11-12 16:41:57 +0000 UTC, 2020-11-12 18:42:07 +0000 UTC)
I1112 22:56:38.814636       1 main.go:218] Error syncing csr csr-8wncm: CSR csr-8wncm creation time 2020-11-12 22:34:47 +0000 UTC not in range (2020-11-12 16:41:57 +0000 UTC, 2020-11-12 18:42:07 +0000 UTC)
```
How is the range determined?  Is it 2 hours from when the cluster is created?  I started the nodes after that so maybe that's the issue?  But that would prevent new nodes from being added automatically...

How are the worker nodes autostarted?

**Comment by fortinj66** [comment]
I found it where the range is determined:

```start := nodeMachine.CreationTimestamp.Add(-maxMachineClockSkew)
	end := nodeMachine.CreationTimestamp.Add(maxMachineDelta)
	if !inTimeSpan(start, end, req.CreationTimestamp.Time) {
		return fmt.Errorf("CSR %s creation time %s not in range (%s, %s)", req.Name, req.CreationTimestamp.Time, start, end)
	}
```
Looks like it is based on node creation time...

and since the nodes were started several hours after they were created the CSR fails.

I'm going to rebuild the cluster tomorrow and hopefully catch these issues sooner...

**Comment by vrutkovs** [comment]
On existing cluster you can remove existing machines (machineset would create new ones) and see if that helps

**Comment by fortinj66** [comment]
That's exactly what I'm testing right now...  That seems to work, although the nodes are not all autostart.  The CSRs are approved as expected though...

**Comment by fortinj66** [comment]
I've continued testing this morning and have reached the following conclusions based on what I have seen over the last several days of testing:

1. a single worker node at 8 Gig is not enough and the installer only starts one node (if it starts one at all).  A single node at 16 Gig seems fine.  I don't understand why only one worker node is started

2) the ignite timeout for the masters seems to be too low.  I've had several instances where the bootstrap takes over 30 seconds to respond to the master nodes for their configs...  If its under 12 secs it seems to be fine.  I'm not sure what the default timeout it.

3) if a provisioned node is not started within 2 hours of its creation, a manual approval of its CSR is required.

4) A nice to have would be for the default image to already have the vmware tools software ( or equiv ) already installed so that things like IP address would be readily available.


Now that I know all the workarounds, I can continue with my testing...

Thanks very much to @vrutkovs for his valuable assistance.  I've learned a lot about OKD V4.  It's so different from V3

--John
